RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,538 @@
+    // =============================================================================
+    /// WARNING: None of the following functions or overloads thereof are part of the
+    /// public API
+    ///
+
+    // Generates a vector of Rademacher entries using the Random123 library
+    template<SignedInteger sint_t = int64_t, typename state_t = RNGState<DefaultRNG>>
+    state_t generate_rademacher_vector_r123(sint_t* buff, int64_t n, state_t &seed_state) {
+        DefaultRNG rng;
+        auto [ctr, key] = seed_state;
+
+        for (int64_t i = 0; i < n; ++i) {
+            typename DefaultRNG::ctr_type r = rng(ctr, key);
+
+            float rand_value = r123::u01fixedpt<float>(r.v[0]);
+
+            buff[i] = rand_value < 0.5 ? -1 : 1;
+
+            ctr.incr();
+        }
+
+        // Return the updated RNGState (with the incremented counter)
+        return state_t {ctr, key};
+    }
+
+    // Catch-all method for applying the diagonal Rademacher
+    // entries in-place to an input matrix, `A`
+    template<typename T, SignedInteger sint_t = int64_t>
+    void apply_diagonal_rademacher(
+                                bool left, // Pre-multiplying?
+                                blas::Layout layout,
+                                int64_t rows,
+                                int64_t cols,
+                                T* A,
+                                sint_t* diag
+                                ) {
+        //TODO: Investigate better schemes for performing the scaling
+        //TODO: Move to `RandBLAS/util.hh`
+        if(left && layout == blas::Layout::ColMajor) {
+            for(int64_t row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row], rows);
+            }
+        }
+        else if(left && layout == blas::Layout::RowMajor) {
+            for(int64_t row = 0; row < rows; row++) {
+                if(diag[row] > 0)
+                    continue;
+                blas::scal(cols, diag[row], &A[row * cols], 1);
+            }
+        }
+        else if(!left && layout == blas::Layout::ColMajor) {
+            for(int64_t col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col * rows], 1);
+            }
+        }
+        else {
+            for(int64_t col=0; col < cols; col++) {
+                if(diag[col] > 0)
+                    continue;
+                blas::scal(rows, diag[col], &A[col], cols);
+            }
+        }
+    }
+
+    template<typename T, SignedInteger sint_t = int64_t>
+    void permute_rows_to_top(
+                          blas::Layout layout,
+                          int64_t rows,
+                          int64_t cols,
+                          sint_t* selected_rows,
+                          int64_t d, // size of `selected_rows`
+                          T* A
+                          ) {
```

Comment:
dimensions:
* m, number of rows in A
* n, number of columns in A
* lda, leading dimension of mat(A) when reading from A in "layout" order.